File: python/HR/Employee.py
from abc import ABC, abstractmethod

class Employee(ABC):

    def __init__(self, id, firstName, lastName, email, contactNumber, location, birthDate):
        self.id = id
        self.firstName = firstName
        self.lastName = lastName
        self.email = email
        self.contactNumber = contactNumber
        self.location = location
        self.birthDate = birthDate

    @abstractmethod
    def doWork(self):
        pass

    @abstractmethod
    def computePay(self):
        pass

    def __str__(self):
        return f"Employee[ID={self.id}, Name={self.firstName} {self.lastName}, Email={self.email}, Contact={self.contactNumber}, Location={self.location}, BirthDate={self.birthDate}]"

File: Javascript/Lab4/HR/Employee.js
//Parent Class

class Employee{
    constructor(id, firstName, lastName, email, contactNumber, location, birthDate){
        if(new.target === Employee){
            throw new Error("Employee is an abstract class and cannot be instantiated");
        }

        this.id = id;
        this.firstName = firstName;
        this.lastName = lastName;
        this.email = email;
        this.contactNumber = contactNumber;
        this.location = location;
        this.birthDate = birthDate;
    }

    doWork(){
        throw new Error("doWork() must be implemented by derivated classes");
    }

    computePay(){
        return 0;
    }

    toString(){
        return `Employee [${this.constructor.name}]
        ID: ${this.id}
        Name: ${this.firstName} ${this.lastName}
        Email: ${this.email}
        Location: ${this.location}`;
    }

}
module.exports = Employee;
